typing/typedecl_unboxed.ml

```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+      pp_shape head_blocks
+
+  let any = { head_imm = Shape_any; head_blocks = Shape_any }
+
+  let none = { head_imm = Shape_set []; head_blocks = Shape_set [] }
+
+  let block_shape tags =
+    { head_imm = Shape_set []; head_blocks = Shape_set tags }
+
+  let cstrs_shape ~num_consts ~num_nonconsts =
+    let int_list n = List.init n Fun.id in
+    {
+      head_imm = Shape_set (int_list num_consts);
+      head_blocks = Shape_set (int_list num_nonconsts)
+    }
+
+  let disjoint_union hd1 hd2 =
+    let union s1 s2 = match s1, s2 with
+    | Shape_any, Shape_set [] | Shape_set [], Shape_any -> Shape_any
+    | Shape_any, _ | _, Shape_any -> raise Conflict
+    | Shape_set l1, Shape_set l2 ->
+        (* invariant : l1 and l2 are sorted *)
+        let rec merge l1 l2 = match l1, l2 with
+        | [], l | l, [] -> l
+        | x :: xx, y :: yy ->
+            if x = y then raise Conflict
+            else if x < y then x :: (merge xx l2)
+            else y :: (merge l1 yy)
+        in Shape_set (merge l1 l2)
+    in
+    {
+      head_imm = union hd1.head_imm hd2.head_imm;
+      head_blocks = union hd1.head_blocks hd2.head_blocks
+    }
+
+  module Callstack = struct
+    type t = Path.t list
+
+    module TypeMap = Btype.TypeMap
+    type map = t TypeMap.t
+
+    let visit p callstack : t =
+      p :: callstack
+
+    let visited p callstack =
+      List.exists (Path.same p) callstack
+
+    let head (callstack_map : map) ty =
+      TypeMap.find ty callstack_map
+
+    let rec extend callstack_map ty new_callstack =
+      let ty = Ctype.repr ty in
+      if TypeMap.mem ty callstack_map then callstack_map
+      else
+        let callstack_map = TypeMap.add ty new_callstack callstack_map in
+        Btype.fold_type_expr (fun callstack_map ty' ->
+          extend callstack_map ty' new_callstack
+        ) callstack_map ty
+
+    let fill ty callstack = extend TypeMap.empty ty callstack
+  end
+
+  (* Useful to interactively debug 'of_type' below. *)
+  let _print_annotations ty callstack_map =
+    Format.eprintf "@[CALLSTACK(%a): @["
+      Printtyp.type_expr ty;
+    let pp_sep ppf () = Format.fprintf ppf ",@ " in
+    Callstack.TypeMap.to_rev_seq callstack_map |> Seq.iter
+      (fun (ty, callstack) ->
+        Format.eprintf "@ @[%a@ [%a]@]"
+          Printtyp.type_expr ty
+          (Format.pp_print_list ~pp_sep Printtyp.path) callstack
+      );
+    Format.eprintf "@]@]@.";
+    ()
+
+  let check_annotated ty callstack =
+    let hash = Btype.TypeHash.create 42 in
+    let rec loop ty =
+      let ty = Ctype.repr ty in
+      if Btype.TypeHash.mem hash ty then ()
+      else begin
+        Btype.TypeHash.add hash ty ();
+        assert (Btype.TypeMap.mem ty callstack);
+        Btype.iter_type_expr loop ty;
+      end
+    in loop ty
+
+  let ty_of_poly ty = match ty.desc with
+  | Tpoly (t, _) -> t
+  | _ -> ty
+
+  let rec of_type env ty callstack_map =
+    let ty = Ctype.repr ty in
+    check_annotated ty callstack_map;
```

Comment:
Note: I still wonder if we could use `Ctype.expand_head_opt` here. (Maybe we didn't try well enough last time?). I think this will be easier to answer (and discuss with more knowledgeable people) in the `get_unboxed_type_repr` PR, assuming that PR gets a better testsuite than we have now. (I have `-dtypedecl-unboxed-repr` somewhere on my TODO list towards that.)